docs/rule_authors/writing_toolchains.md

```diff
@@ -0,0 +1,130 @@
+---
+id: writing_toolchains
+title: Writing Toolchains
+---
+
+Toolchains are regular rules that:
+
+- Have `is_toolchain_rule = True` passed to the `rule` call.
+- Return a toolchain provider, which is just a struct with a name like
+  `*ToolchainInfo`.
+
+Toolchain rules are instantiated once in the `toolchains//` cell. The
+location of the `toolchains` cell is determined by the value of
+`cells.toolchains` in the `.buckconfig` file.
+
+Regular build rules reference those toolchain targets as
+[`toolchain_dep`](https://buck2.build/docs/api/build/attrs/#toolchain_dep)
+attrs (often
+[`default_only`](https://buck2.build/docs/api/build/attrs/#default_only)
+ones). See
+<https://buck2.build/docs/rule_authors/configurations/#toolchain-deps>
+for more information about toolchain deps.
+
+## Writing a custom toolchain
+
+People will often first encounter toolchains when they want to switch
+off of the demo toolchains that `buck2 init` uses by default. For
+example, one might want to tweak which compiler is used, which flags are
+passed to it, or where it is fetched from.
+
+The most straightforward thing to do first is to instantiate one of the
+"system" toolchain rules that the prelude offers. Studying the
+toolchains defined by the `system_demo_toolchains` macro in
+`@prelude//toolchains:demo.bzl` is a good way to get started.
+
+For example, here is how one could define a C++ toolchain that builds
+projects in C++23, with warnings as errors and optimizations enabled:
+
+```python
+load("@prelude//toolchains:cxx.bzl", "system_cxx_toolchain")
+
+system_cxx_toolchain(
+    name = "cxx",
+    compiler_type = "clang",
+    cxx_flags = [
+        "-std=c++23",
+        "-Wall",
+        "-Wextra",
+        "-Werror",
+        "-O3",
+    ],
+    visibility = ["PUBLIC"],
+)
+```
+
+One would typically use
+[`select`](https://buck2.build/docs/rule_authors/configurations_by_example/)
+to customize the toolchain e.g. based on the build mode (debug vs
+release) or compiler type.
+
+Note that several toolchains require you to also define a
+`python_bootstrap` toolchain. This is because those toolchains use
+Python scripts (e.g. to generate compilation databases for C++), and the
+prelude gives you control over the Python toolchain used for this
+purpose. The bootstrap Python toolchain can be different from the
+toolchain used by `python_*` rules. `@prelude//toolchains:python.bzl`
+provides a simple `system_python_bootstrap_toolchain` in case you do not
+care about tightly controlling the Python used when building other
+languages.
+
+Completely custom toolchains can also easily be written as rules with
+`is_toolchain_rule = True` which return a `*ToolchainInfo` struct. The
```

Comment:
"completely custom" (e.g. not a language/process known to the prelude) seems to be at odds with "use one of the pre-defined `*ToolchainInfo` structs". What if I want a toolchain for something the prelude doesn't have a pre-defined helper for? Clarify and contrast with e.g. `RunInfo`?

Reply:
Good point, I think it's worth clarifying a few different use cases:

- Defining a custom toolchain for a language that is supported in the prelude, in which case one could either:
    - Return the `*ToolchainInfo` provider defined in the prelude, to get all the prelude target rules for free.
    - Return a custom toolchain provider, for use with a custom set of target rules (which is a lot more effort).
- Defining a toolchain for a custom language/process, in which case one would have to define a toolchain provider for it